src/snac_model.cpp

```diff
@@ -0,0 +1,220 @@
+    }
+    return embd;
+}
+
+snac_context * build_new_snac_context(struct snac_model * model, int n_threads, bool use_cpu) {
+    snac_context * sctx = new snac_context(model, n_threads);
+    if (!use_cpu) {
+#ifdef GGML_USE_METAL
+        sctx->backend = ggml_backend_metal_init();
+#endif
+    }
+    sctx->backend_cpu = ggml_backend_cpu_init();
+    sctx->set_threads();
+    sctx->build_schedule();
+    sctx->buf_compute_meta.resize(ggml_tensor_overhead()*model->max_nodes() + ggml_graph_overhead_custom(model->max_nodes(), false));
+    return sctx;
+}
+
+void snac_runner::prepare_post_load() {
+    ggml_cgraph * gf = build_snac_graph(model->max_generation_size);
+    sctx->prep_schedule(gf);
+}
+    
+struct ggml_cgraph * snac_runner::build_snac_graph(size_t sequence_length) {
+    init_build();
+    struct ggml_cgraph * gf = ggml_new_graph_custom(ctx, 8192, false);
+    
+    struct ggml_tensor * cur;
+    struct ggml_tensor * inputs;
+
+    if (model->use_noise) {
+        sctx->noise = ggml_new_tensor_1d(ctx, GGML_TYPE_F32, model->noise_steps_sum * sequence_length);
+        ggml_set_input(sctx->noise);
+    }
+    
+    inputs = snac_build_audio_inputs(ctx, sctx, sequence_length, model->quantizer_layers);
+    cur = ggml_conv_1d_dw(ctx, model->in_conv_kernel, inputs, 1, 3, 1);
+    cur = ggml_add(ctx, cur, model->in_conv_bias);
+    cur = ggml_conv_1d(ctx, model->up_conv_kernel, cur, 1, 0, 1);
+    cur = ggml_add(ctx, cur, model->up_conv_bias);
+    size_t noise_offset = 0;
+    for (int l = 0; l < model->layers.size(); l++) {
+        auto layer = model->layers[l]; 
+        struct ggml_tensor * noise = nullptr;
+        if (model->use_noise) {
+            noise = ggml_cont(ctx, ggml_view_1d(ctx, sctx->noise, model->noise_steps[l] * sequence_length, noise_offset));
+            noise_offset += model->noise_steps[l] * sequence_length * sizeof(float);
+        }
+        cur = general_neural_audio_codec::build_layer(ctx, cur, layer, noise);
+    }
+    cur = snake_1d(ctx, model->snake_alpha, cur);
+    cur = ggml_conv_1d(ctx, model->out_conv_kernel, cur, 1, 3, 1);
+    cur = ggml_add(ctx, cur, model->out_conv_bias);
+    cur = ggml_tanh(ctx, cur);
+    ggml_build_forward_expand(gf, cur);
+    free_build();
+    return gf;
+}
+
+void snac_runner::set_inputs(std::vector<std::vector<uint32_t>> & tokens) {
+    ggml_backend_tensor_set(
+        sctx->inp_tokens, tokens[0].data(), 0, 
+        tokens[0].size()*ggml_element_size(sctx->inp_tokens)
+    );
+
+    ggml_backend_tensor_set(
+        sctx->inp_tokens, tokens[1].data(), tokens[0].size() * ggml_element_size(sctx->inp_tokens), 
+        tokens[1].size() * ggml_element_size(sctx->inp_tokens)
+    );
+
+    ggml_backend_tensor_set(
+        sctx->inp_tokens, tokens[2].data(), 
+        tokens[1].size()*ggml_element_size(sctx->inp_tokens)+tokens[0].size()*ggml_element_size(sctx->inp_tokens), 
+        tokens[2].size()*ggml_element_size(sctx->inp_tokens)
+    );
+    if (model->use_noise) {
```

Comment:
Currently, we are always adding noise, so I should probably remove this. Spot checking showed no difference in outcome when we don't add noise, but the performance improvement is only ~1 ms.